src/chips/sam4l/spi_dma.rs

```diff
@@ -0,0 +1,383 @@
+use dma::DMAChannel;
+use dma::DMAClient;
+use gpio::PC;
+
+/// Implementation of DMA-based SPI master communication for
+/// the Atmel SAM4L CortexM4 microcontroller.
+/// Authors: Sam Crow <samcrow@uw.edu>
+///          Philip Levis <pal@cs.stanfored.edu>
+///
+// Driver for the SPI hardware (seperate from the USARTS),
+// described in chapter 26 of the datasheet
+
+/// The registers used to interface with the hardware
+#[repr(C, packed)]
+struct SpiRegisters {
+    cr: u32, // 0x0
+    mr: u32, // 0x4
+    rdr: u32, // 0x8
+    tdr: u32, // 0xC
+    sr: u32, // 0x10
+    ier: u32, // 0x14
+    idr: u32, // 0x18
+    imr: u32, // 0x1C
+    reserved0: [u32; 4], // 0x20, 0x24, 0x28, 0x2C
+    csr0: u32, // 0x30
+    csr1: u32, // 0x34
+    csr2: u32, // 0x38
+    csr3: u32, // 0x3C
+    reserved1: [u32; 41], // 0x40 - 0xE0
+    wpcr: u32, // 0xE4
+    wpsr: u32, // 0xE8
+    reserved2: [u32; 3], // 0xEC - 0xF4
+    features: u32, // 0xF8
+    version: u32, // 0xFC
+}
+
+const SPI_BASE: u32 = 0x40008000;
+
+/// Values for selected peripherals
+#[derive(Copy,Clone)]
+pub enum Peripheral {
+    Peripheral0,
+    Peripheral1,
+    Peripheral2,
+    Peripheral3,
+}
+
+///
+/// The SAM4L supports four peripherals. 
+pub struct Spi {
+    regs: *mut SpiRegisters,
+    callback: Option<&'static SpiCallback>,
+    dma_read:  Option<&'static mut DMAChannel>,
+    dma_write: Option<&'static mut DMAChannel>,
+    // keep track of which // interrupts are pending in order 
+    // to correctly issue completion event only after both complete.
+    reading: Cell<bool>,
+    writing: Cell<bool>,
+}
+
+pub static mut SPI: Spi = Spi::new();
+
+impl Spi {
+   /// Creates a new SPI object, with peripheral 0 selected
+   pub const fn new() -> Spi {
+        Spi {
+            regs: SPI_BASE as *mut SpiRegisters,
+            callback: None,
+            dma_read:  None,
+            dma_write: None,
+            reading: Cell::new(false),
+            writing: Cell::new(false)
+        }
+    }
+
+    pub fn enable(&self) {
+        self.dma_read.as_ref().map(|read| read.enable());
+        self.dma_write.as_ref().map(|write| write.enable());
+        unsafe { volatile_store(&mut (*self.regs).cr, 0b1); }
+    }
+
+    pub fn disable(&self) {
+        self.dma_read.as_ref().map(|read| read.disable());
+        self.dma_write.as_ref().map(|write| write.disable());
+        unsafe { volatile_store(&mut (*self.regs).cr, 0b10); }
+    }
+
+    /// Sets the approximate baud rate for the active peripheral,
+    /// and return the actual baud rate set.
+    ///
+    /// Since the only supported baud rates are 48 MHz / n where n 
+    /// is an integer from 1 to 255, the exact baud rate may not 
+    /// be available. In that case, the next lower baud rate will 
+    /// be selected.
+    ///
+    /// The lowest available baud rate is 188235 baud. If the 
+    /// requested rate is lower, 188235 baud will be selected.
+    pub fn set_baud_rate(&self, rate: u32) -> u32 {
```

Comment:
I wonder if this method should take in only acceptable values as arguments (e.g. `n: u8`) and the logic for approximating a good value for `n` based on a target baud rate should be in the implementation of `SpiMaster` trait's `set_rate` method.